### PR TITLE
Add branch condition for calling move to device in prefetch (FSDP 3/n)

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Adds an additional property to the DDP/DDP Spawn Plugin that allows plugins that extend DDP to disable moving model to device. In FSDP this is required since we can only move to device after configure_ddp is called, which would require overriding this function pre_dispatch. Also fixed an issue where spawning the DDP Plugin was not possible because we missed copying a default setter.

Fixes #<issue_number>

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/accelerators/test_ddp.py [hunk 1]
```diff
@@ -102,3 +103,28 @@ def test_torch_distributed_backend_env_variables(tmpdir):
                 logger=False,
             )
             trainer.fit(model)
+
+
+@pytest.mark.parametrize('move_to_device_pre_dispatch_enabled', [True, False])
+@mock.patch('pytorch_lightning.plugins.DDPPlugin.model_to_device')
+def test_move_to_device_in_pre_dispatch(mock_model_to_device, tmpdir, move_to_device_pre_dispatch_enabled):
```
Comment (SeanNaren) [inline tests/accelerators/test_ddp.py:110]: Two things:

- Would've liked to add a spawn test but mocks don't seem to carry across to new processes?
- Can I combine patch and parametrize?

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 2]
```diff
@@ -59,7 +59,7 @@ def __init__(
         self.sync_batchnorm = sync_batchnorm
         self._ddp_kwargs = kwargs
         self.dist = LightningDistributed()
-        self.num_processes = len(parallel_devices)
+        self.num_processes = len(parallel_devices) if parallel_devices is not None else parallel_devices
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:62]:
```suggestion
        self.num_processes = len(parallel_devices) if parallel_devices is not None else None
```

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 3]
```diff
@@ -313,3 +313,11 @@ def predict(self, *args, **kwargs):
...
         if not self.lightning_module.automatic_optimization:
             self.model.require_backward_grad_sync = True
+
+    @property
+    def call_move_to_device_hook_in_pre_dispatch(self) -> bool:
+        """
+        Call the ``model_to_device`` function within pre_dispatch if this is set to True.
+        Useful for when plugin would like to call model_to_device at another time, or skip the call.
+        """
+        return True
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp.py:323]: Should this be in `ParallelPlugin`?

File: tests/accelerators/test_ddp.py [hunk 4]
```diff
@@ -102,3 +103,28 @@ def test_torch_distributed_backend_env_variables(tmpdir):
...
+    in training.
+    """
+
+    with mock.patch(
+        f'pytorch_lightning.plugins.DDPPlugin.call_move_to_device_hook_in_pre_dispatch',
+        move_to_device_pre_dispatch_enabled
+    ):
+        model = BoringModel()
+        trainer = Trainer(
+            default_root_dir=tmpdir, fast_dev_run=True, accelerator='ddp', plugins=DDPPlugin(), num_processes=1
```
Comment (carmocca) [inline tests/accelerators/test_ddp.py:122]: Why do you need to pass `DDPPlugin`?

File: tests/accelerators/test_ddp.py (hunk 1 above)
Comment (carmocca) [reply tests/accelerators/test_ddp.py:110]: 1. Nope, you'd have to create the mock in each process
2. What do you mean? Having different mocks for each parametrize? I don't think so

Comment (tchaton) [reply tests/accelerators/test_ddp.py:110]: You might be able to do so with a callback, applying the patch context manager.

Comment (tchaton) [reply tests/accelerators/test_ddp.py:110]: You can make patch conditional on parametrize. with the return object.

**Comment by tchaton** [comment]
Hey @SeanNaren. Any updates ?

**Comment by SeanNaren** [comment]
Given how little now exists in the pre_dispatch function after #6506, we don't need this anymore :)
